deriveLift-generated instance for unlifted newtype fails to typecheck

The following should work on GHC 8.10 or later, but doesn't:

```hs
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UnliftedNewtypes #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Bug where

import GHC.Exts
import Language.Haskell.TH.Lift

newtype T = MkT Int#
$(deriveLift ''T)
```
```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )
Bug.hs:11:3-16: Splicing declarations
    deriveLift ''T
  ======>
    instance Lift T where
      lift
        = \ e_a4bz
            -> case e_a4bz of {
                 MkT x0_a4bA
                   -> (Language.Haskell.TH.Lib.Internal.appE
                         (Language.Haskell.TH.Lib.Internal.conE
                            ((Language.Haskell.TH.Syntax.Name
                                (Language.Haskell.TH.Syntax.mkOccName "MkT"))
                               (((Language.Haskell.TH.Syntax.NameG
                                    Language.Haskell.TH.Syntax.DataName)
                                   (Language.Haskell.TH.Syntax.mkPkgName "fake-package-0-inplace"))
                                  (Language.Haskell.TH.Syntax.mkModName "Bug")))))
                        (Language.Haskell.TH.Lib.Internal.litE
                           (Language.Haskell.TH.Lib.Internal.intPrimL
                              (toInteger (I# x0_a4bA)))) }
      liftTyped = (Language.Haskell.TH.Syntax.unsafeTExpCoerce . lift)

Bug.hs:11:3: error:
    • Couldn't match a lifted type with an unlifted type
      When matching types
        t0 :: *
        T :: TYPE 'IntRep
      Expected type: T
                     -> Language.Haskell.TH.Syntax.Q (Language.Haskell.TH.Syntax.TExp T)
        Actual type: t0
                     -> Language.Haskell.TH.Syntax.Q (Language.Haskell.TH.Syntax.TExp T)
    • In the expression:
        (Language.Haskell.TH.Syntax.unsafeTExpCoerce . lift)
      In an equation for ‘liftTyped’:
          liftTyped = (Language.Haskell.TH.Syntax.unsafeTExpCoerce . lift)
      In the instance declaration for ‘Lift T’
   |
11 | $(deriveLift ''T)
   |   ^^^^^^^^^^^^^^
```

The issue is rather silly: the use of `(.)` in `liftTyped = unsafeTExpCoerce . lift` forces the types to be lifted instead of using `TYPE 'IntRep`. Simply generating `liftTyped = \x -> unsafeTExpCoerce (lift x)` instead would fix the issue.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,11 @@
 
 All notable changes to this project will be documented in this file.
 
+## [next] - ????-??-??
+
+* Allow building with `template-haskell-2.17.0.0` (GHC 9.0).
+* Make `deriveLift` work for unlifted newtypes.
+
 ## [0.8.1] - 2019-12-06
 
 * Support GHC 8.10/`template-haskell-2.16`.

--- a/src/Language/Haskell/TH/Lift.hs
+++ b/src/Language/Haskell/TH/Lift.hs
@@ -34,6 +34,7 @@ import Data.Char (ord)
 import Language.Haskell.TH
 import Language.Haskell.TH.Datatype
 import qualified Language.Haskell.TH.Lib as Lib (starK)
+import Language.Haskell.TH.Lift.Internal
 import Language.Haskell.TH.Syntax
 import Control.Monad ((<=<), zipWithM)
 #if MIN_VERSION_template_haskell(2,9,0)
@@ -140,17 +141,19 @@ deriveLiftOne i = withInfo i liftInstance
 #else /* MIN_VERSION_template_haskell(2,9,0) */
       let phtys = []
 #endif
+      _x <- newName "x"
       instanceD (ctxt dcx phtys tys)
                 (conT ''Lift `appT` typ n tys)
                 [ funD 'lift [clause [] (normalB (makeLiftOne n cons)) []]
 #if MIN_VERSION_template_haskell(2,16,0)
-                , funD 'liftTyped [clause [] (normalB [| unsafeTExpCoerce . lift |]) []]
+                , let rhs = varE 'unsafeSpliceCoerce `appE`
+                              (varE 'lift `appE` varE _x) in
+                  funD 'liftTyped [clause [varP _x] (normalB rhs) []]
 #endif
                 ]
     typ n = foldl appT (conT n) . map unKind
-    -- Only consider *-kinded type variables, because Lift instances cannot
-    -- meaningfully be given to types of other kinds. Further, filter out type
-    -- variables that are obviously phantom.
+    -- Only consider *-kinded type variables for now. Furthermore, filter out
+    -- type variables that are obviously phantom.
     ctxt dcx phtys =
         fmap (dcx ++) . cxt . concatMap liftPred . filter (`notElem` phtys)
     liftPred ty =
@@ -176,7 +179,8 @@ makeLiftOne n cons = do
 consMatches :: Name -> [ConstructorInfo] -> [Q Match]
 consMatches n [] = [match wildP (normalB e) []]
   where
-    e = varE 'errorQExp `appE` (stringE $ "Can't lift value of empty datatype " ++ nameBase n)
+    e = varE 'errorQuoteExp `appE`
+             (stringE $ "Can't lift value of empty datatype " ++ nameBase n)
 consMatches _ cons = concatMap doCons cons
 
 doCons :: ConstructorInfo -> [Q Match]
@@ -247,34 +251,28 @@ withInfo i f = case i of
                  } ->
       f dcx n vs cons
 
--- A type-restricted version of error that ensures makeLift always returns a
--- value of type Q Exp, even when used on an empty datatype.
-errorQExp :: String -> Q Exp
-errorQExp = error
-{-# INLINE errorQExp #-}
-
 instance Lift Name where
   lift (Name occName nameFlavour) = [| Name occName nameFlavour |]
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 instance Lift OccName where
   lift n = [| mkOccName |] `appE` lift (occString n)
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 instance Lift PkgName where
   lift n = [| mkPkgName |] `appE` lift (pkgString n)
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 instance Lift ModName where
   lift n = [| mkModName |] `appE` lift (modString n)
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 instance Lift NameFlavour where
@@ -292,7 +290,7 @@ instance Lift NameFlavour where
   lift (NameG nameSpace' pkgName modnam)
    = [| NameG nameSpace' pkgName modnam |]
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 instance Lift NameSpace where
@@ -300,5 +298,5 @@ instance Lift NameSpace where
   lift DataName = [| DataName |]
   lift TcClsName = [| TcClsName |]
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif

--- a/src/Language/Haskell/TH/Lift/Internal.hs
+++ b/src/Language/Haskell/TH/Lift/Internal.hs
@@ -1,0 +1,44 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+#if __GLASGOW_HASKELL__ >= 706
+{-# LANGUAGE PolyKinds #-}
+#endif
+
+-- | Helper functions used in code that "Language.Haskell.TH.Lift" generates.
+--
+-- Note: this is an internal module, and as such, the API presented here is not
+-- guaranteed to be stable, even between minor releases of this library.
+module Language.Haskell.TH.Lift.Internal where
+
+#if MIN_VERSION_template_haskell(2,16,0)
+import GHC.Exts (RuntimeRep, TYPE)
+#endif
+
+import Language.Haskell.TH.Syntax
+
+-- | A type-restricted version of 'error' that ensures 'makeLift' always
+-- returns a value of type @q 'Exp'@ (where @q@ is an instance of 'Quote'),
+-- even when used on an empty datatype.
+#if MIN_VERSION_template_haskell(2,17,0)
+errorQuoteExp :: Quote q => String -> q Exp
+#else
+errorQuoteExp ::            String -> Q Exp
+#endif
+errorQuoteExp = error
+
+-- | This is a cargo-culted version of @unsafeSpliceCoerce@ from the
+-- @th-compat@ library, which has been copied here to avoid incurring a library
+-- dependency.
+--
+-- Only available when built with @template-haskell-2.9.0.0@ or later.
+#if MIN_VERSION_template_haskell(2,17,0)
+unsafeSpliceCoerce :: forall (r :: RuntimeRep) (a :: TYPE r) m. Quote m => m Exp -> Code m a
+unsafeSpliceCoerce = unsafeCodeCoerce
+#elif MIN_VERSION_template_haskell(2,16,0)
+unsafeSpliceCoerce :: forall (r :: RuntimeRep) (a :: TYPE r). Q Exp -> Q (TExp a)
+unsafeSpliceCoerce = unsafeTExpCoerce
+#elif MIN_VERSION_template_haskell(2,9,0)
+unsafeSpliceCoerce :: forall a. Q Exp -> Q (TExp a)
+unsafeSpliceCoerce = unsafeTExpCoerce
+#endif

--- a/t/Foo.hs
+++ b/t/Foo.hs
@@ -7,6 +7,10 @@
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
+
+#if __GLASGOW_HASKELL__ >= 810
+{-# LANGUAGE UnliftedNewtypes #-}
+#endif
 module Foo where
 
 import GHC.Prim (Double#, Float#, Int#, Word#)
@@ -16,7 +20,7 @@ import GHC.Prim (Char#)
 
 import Language.Haskell.TH.Lift
 #if MIN_VERSION_template_haskell(2,16,0)
-import Language.Haskell.TH.Syntax (unsafeTExpCoerce)
+import Language.Haskell.TH.Lift.Internal (unsafeSpliceCoerce)
 #endif
 
 -- Phantom type parameters can't be dealt with poperly on GHC < 7.8.
@@ -65,7 +69,7 @@ $(deriveLift ''Unboxed)
 instance Lift (f (Fix f)) => Lift (Fix f) where
   lift = $(makeLift ''Fix)
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 
 #if MIN_VERSION_template_haskell(2,7,0)
@@ -73,7 +77,7 @@ $(deriveLift 'FamPrefix1)
 instance (Eq a, Lift a) => Lift (Fam a Bool Bool) where
   lift = $(makeLift 'FamInstBool)
 #if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped = unsafeTExpCoerce . lift
+  liftTyped = unsafeSpliceCoerce . lift
 #endif
 #endif
 
@@ -97,9 +101,15 @@ data instance Fam2 a Bool Bool = Fam2InstBool a Bool
 $(pure [])
 
 instance Lift (f (Fix2 f)) => Lift (Fix2 f) where
-  liftTyped = unsafeTExpCoerce . $(makeLift ''Fix2)
+  liftTyped = unsafeSpliceCoerce . $(makeLift ''Fix2)
 instance Lift a => Lift (Fam2 a Int Char) where
-  liftTyped = unsafeTExpCoerce . $(makeLift 'Fam2Prefix1)
+  liftTyped = unsafeSpliceCoerce . $(makeLift 'Fam2Prefix1)
 instance (Eq a, Lift a) => Lift (Fam2 a Bool Bool) where
-  liftTyped = unsafeTExpCoerce . $(makeLift 'Fam2InstBool)
+  liftTyped = unsafeSpliceCoerce . $(makeLift 'Fam2InstBool)
+#endif
+
+#if __GLASGOW_HASKELL__ >= 810
+-- Regression test for #43
+newtype T43 = MkT43 Int#
+$(deriveLift ''T43)
 #endif

--- a/th-lift.cabal
+++ b/th-lift.cabal
@@ -37,6 +37,7 @@ source-repository head
 Library
   Default-Language: Haskell2010
   Exposed-modules:  Language.Haskell.TH.Lift
+                    Language.Haskell.TH.Lift.Internal
   Other-Extensions: CPP,  MagicHash, TypeSynonymInstances, FlexibleInstances
   if impl(ghc >= 8.0)
     Other-Extensions: TemplateHaskellQuotes
@@ -45,8 +46,8 @@ Library
   Hs-Source-Dirs:   src
   Build-Depends:    base              >= 4.3  && < 5,
                     ghc-prim,
-                    th-abstraction   >= 0.2.3 && < 0.4,
-                    template-haskell >= 2.5   && < 2.17
+                    th-abstraction   >= 0.2.3 && < 0.5,
+                    template-haskell >= 2.5   && < 2.18
   ghc-options:      -Wall
 
 Test-Suite test

